### ISSUE TITLE
Users cannot break out of inline nodes at the end of an editor due to #4578

### ISSUE THREAD

**Issue by AleksandrHovhannisyan** [ianstormtaylor/slate#4704]
## Description

#4578 changed the behavior for how typing works at the end of inline nodes, allowing users to continue typing while their cursor is on either side of an inline node (e.g., a hyperlink). The PR fixed that problem, but it introduced a related problem: Now, users are unable to break out of inline nodes using arrow keys at the end of the editor. It's worth noting that this change was not documented; as far as I can tell, it was not mentioned in Releases and presumably went in as a patch or minor upgrade (0.67 maybe?).

### Current behavior is unintuitive

We did some internal testing, and non-technical users who are unfamiliar with this change could not figure out how to break out of inline nodes at the end of the editor. This is actually a fairly common situation—imagine you are authoring content and you want to insert a link. If you do this as you are typing, you won't be able to continue typing at the end of the editor.

Arguably, this creates a more frustrating end-user experience than if users cannot continue typing at the end of an inline node (the original bug). The latter can easily be fixed by backtracking the cursor, inserting text in the middle of the inline, and forward-deleting the unwanted content. It's not ideal, but at least your cursor is not stuck in the inline anymore.

### The proposed solution is also unintuitive

[There was some discussion about this change in that PR](https://github.com/ianstormtaylor/slate/pull/4578#issuecomment-947643779); the recommendation was for Slate developers to alter their implementation to change how the left and right arrow keys behave. Assuming that Slate users do implement this proposed change, it does not fix the underlying UX problem. From an end user's perspective, it is not intuitive to have to press the left/right arrow keys _twice_ to be able to enter and exit inline nodes, as demonstrated in https://www.slatejs.org/examples/inlines. Users expect that pressing the left arrow key moves them one unit to the left, and pressing the right arrow key moves them one unit to the right, like in most other text editors. 

Implementing this new behavior requires developers to alter the design for inline nodes to provide some sort of visual indication that the cursor is actively focusing an inline node vs. "sitting right next to an inline but technically outside it." Without a visual indicator, users may think this is a bug because the cursor appears to be stuck in a single location and doesn't move. Slate should try to maintain sensible default behaviors that don't require developers to alter any designs. We are effectively hijacking how the left/right arrow keys behave. This may even create accessibility issues.

## Recording

![demo](https://user-images.githubusercontent.com/19352442/144305535-3775f8e3-00df-4540-b2f4-95bc0d38aea6.gif)

## Sandbox

https://codesandbox.io/s/slate-reproductions-forked-pc1g9?file=/index.js

## Steps to reproduce
1. Create an editor whose last child is an inline.
2. Put your cursor at the end of the inline.
3. Start typing.
4. Observe that you cannot break out of the inline.

## Expectation

The user experience needs to be reconsidered for how typing works at the end of an inline node. What this entails is unclear. It may require reverting #4578. In the meantime, our team may need to downgrade Slate.

## Environment
- Slate Version: 0.70
- Operating System: Tested on Windows and Mac
- Browser: all

**Comment by jameshfisher** [comment]
(Author of https://github.com/ianstormtaylor/slate/pull/4578 here)

I agree that this is a problem, because it's one that I have as well!

I would strongly feel that reverting https://github.com/ianstormtaylor/slate/pull/4578 would be the wrong solution, though, because it's easy to _add_ the "jump out of the inline" feature if you want it, but hard/impossible to _remove_ the "jump out of the inline" feature if you don't want it.

I could contribute a tiny plugin like `withInlineEscaper`, that basically just re-adds that behavior. @AleksandrHovhannisyan would that be appreciated?

(However, I think the _better_ approach would be to solve the underlying bug: when you click at the end of the paragraph, it should drop the cursor into the end of the empty `{text: ''}` at the end of the paragraph (which should always exist due to normalization rules). Currently, it instead seems to drop the cursor at the end of the inline element.)

**Comment by jameshfisher** [comment]
I believe it's a browser bug that causes the bad "click at end of line" behavior. I found a possible fix/workaround: make the text element non-zero-width, e.g. add `padding-right: 0.001em` to it. The browser (at least, Chrome) then stops ignoring it when deciding where to place the cursor when a click happens. @AleksandrHovhannisyan here's a POC using your example: https://codesandbox.io/s/fix-click-at-end-of-line-by-adding-padding-to-text-bkfw9?file=/components/CustomText.jsx

**Comment by jameshfisher** [comment]
It's certainly a browser bug, because Firefox has different behavior. Using [@AleksandrHovhannisyan's example](https://codesandbox.io/s/slate-reproductions-forked-pc1g9?file=/index.js) on Firefox (v85 on Windows), clicking at the end of the paragraph puts the cursor _outside_ the inline, as desired.

(However, this is slightly obscured by the example, where the starting document is technically invalid. It should have `{ text: '' }` as its last child to satisfy normalization rules.)

**Comment by jameshfisher** [comment]
Apart from clicking at the end of the paragraph, the other thing that users try (@AleksandrHovhannisyan could you confirm?) is to hit the right arrow key. Currently, Slate seems to do nothing here, and I think that's a bug. The current Slate logic is here: https://github.com/ianstormtaylor/slate/blob/e9a46ad29e0376a45051c4a8100c5678784b785c/packages/slate-react/src/components/editable.tsx#L1109-L1131

**Comment by AleksandrHovhannisyan** [comment]
> (However, I think the better approach would be to solve the underlying bug: when you click at the end of the paragraph, it should drop the cursor into the end of the empty {text: ''} at the end of the paragraph (which should always exist due to normalization rules). Currently, it instead seems to drop the cursor at the end of the inline element.)

Yup, agreed! Reverting your PR would just regress to the other bug, which would just make things worse. The ideal fix would be to keep that change but also improve the click-at-end UX, as you mentioned.

> the other thing that users try (@AleksandrHovhannisyan could you confirm?) is to hit the right arrow key.

Yup—we ended up implementing your solution in https://github.com/ianstormtaylor/slate/commit/f1b7d18f43913474617df02f747afa0e78154d85#diff-98b609559db12cb7cc755fd41741b9c72affed934a10cd75db684d5455f3868cR65 and https://github.com/ianstormtaylor/slate/pull/4630 for the time being, and it seems to work well. We had to add a focus ring indicator to make the UX clearer, though.

Regarding the padding trick: That's interesting! I was able to confirm it fixes the click-at-end behavior for Chrome, Firefox, and Safari (both on Windows and mac).

**Comment by jameshfisher** [comment]
> Yup—we ended up implementing your solution in [f1b7d18#diff-98b609559db12cb7cc755fd41741b9c72affed934a10cd75db684d5455f3868cR65](https://github.com/ianstormtaylor/slate/commit/f1b7d18f43913474617df02f747afa0e78154d85#diff-98b609559db12cb7cc755fd41741b9c72affed934a10cd75db684d5455f3868cR65) and #4630 for the time being, and it seems to work well. We had to add a focus ring indicator to make the UX clearer, though.

Okay, great! Yeah I've also used the focus ring technique.

I've added this padding workaround to the standard Slate example in that PR ^^ I would like to better understand the root cause of the bug, but I think this will have to do for now.

I think the remaining issue is the Slate behavior when hitting the right arrow key.

**Comment by lworkman** [comment]
For anyone else running into this, a style rule that only applies the spacing when the last node is empty appears to work. That way you don't get extra spacing between words or letters. Something like:
```
[data-slate-editor="true"] > *:last-child > *:last-child [data-slate-length='0']
```
